jena-arq/src/main/java/org/apache/jena/sparql/engine/main/StageGeneratorGeneric.java

```diff
@@ -49,12 +53,29 @@ public QueryIterator execute(BasicPattern pattern, QueryIterator input, Executio
         return execute(pattern, reorder, input, execCxt) ;
     }
 
+    protected static <T> T closeOnException(Supplier<T> compute, Closeable closeAction) {
+        try {
+            T result = compute.get();
+            return result;
+        } catch (Exception e) {
+            closeAction.close();
+            e.addSuppressed(new RuntimeException("Exception during construction of basic pattern stage execution."));
+            throw e;
+        }
+    }
+
+    /**
+     * Attempts to construct an iterator that executes the input against the pattern.
+     * If the construction fails, such as due to {@link QueryCancelledException}, then the exception is passed on
+     * and the input iterator is closed.
+     */
     protected QueryIterator execute(BasicPattern pattern, ReorderTransformation reorder,
                                     QueryIterator input, ExecutionContext execCxt) {
         Explain.explain(pattern, execCxt.getContext()) ;
 
-        if ( ! input.hasNext() )
-            return input ;
+        QueryIterator originalInput = input;
+        if ( closeOnException(() -> !originalInput.hasNext(), originalInput) )
+            return originalInput ;
```

Comment:
The `input` should be closed in a cancel. If it is not then something is missing elsewhere. 

See the issue #2881 for discussion about inverting the overall design to have explicit declaration and management of resources that have clear-up.

Reply:
Right, I think this check is superfluous.

Reply:
Removed the check again.